text/3519-arbitrary-self-types-v2.md

```diff
@@ -0,0 +1,403 @@
+
+This RFC does not propose any changes to `DispatchFromDyn`. Since `DispatchFromDyn` is unstable at the moment, object-safe receivers might be delayed until `DispatchFromDyn` is stabilized. `Receiver` is not blocked on further `DispatchFromDyn` work, since non-object-safe receivers already cover a big chunk of the use-cases.
+
+## Lifetime elision
+
+As discussed in the [motivation](#motivation), this new facility is _most likely_ to be used in cases where a standard reference can't normally be used. But in other cases a smart pointer self type might wrap a standard Rust reference, and thus might be parameterized by a lifetime.
+
+Lifetime elision works in the expected fashion:
+
+```rust
+struct SmartPtr<'a, T: ?Sized>(&'a T);
+
+impl<'a, T: ?Sized> Receiver for SmartPtr<'a, T> {
+    type Target = T;
+}
+
+struct MyType;
+
+impl MyType {
+    fn m(self: SmartPtr<Self>) {}
+    fn n(self: SmartPtr<'_, Self>) {}
+    fn o<'a>(self: SmartPtr<'a, Self>) {}
+}
+```
+
+## Diagnostics
+
+The existing branches in the compiler for "arbitrary self types" already emit excellent diagnostics. We will largely re-use them, with the following improvements:
+
+- In the case where a self type is invalid because it doesn't implement `Receiver`, the existing excellent error message will be updated.
+- An easy mistake is to implement `Receiver` for `P<T>`, forgetting to specify `T: ?Sized`. `P<Self>` then only works as a `self` parameter in traits `where Self: Sized`, an unusual stipulation. It's not obvious that `Sized`ness is the problem here, so we will identify this case specifically and produce an error giving that hint.
+- There are certain types which feel like they "should" implement `Receiver` but do not: `Weak` and `NotNull`. If these are encountered as a self type, we should produce a specific diagnostic explaining that they do not implement `Receiver` and suggesting that they could be wrapped in a newtype wrapper if method calls are important. This will require `Weak` and `NonNull` be marked as lang items so that the compiler is aware of the special nature of these types. (The authors of this RFC feel that these extra lang-items _are_ worthwhile to produce these improved diagnostics - if the reader disagrees, please let us know.)
+- Under some circumstances, the compiler identifies method candidates but then discovers that the self type doesn't match. This results currently in a simple "mismatched types" error; we can provide a more specific error message here. The only known case is where a method is generic over `Receiver`, and the caller explicitly specifies the wrong type:
+    ```rust
+    #![feature(receiver_trait)]
+
+    use std::ops::Receiver;
+
+    struct SmartPtr<'a, T: ?Sized>(&'a T);
+
+    impl<'a, T: ?Sized> Receiver for SmartPtr<'a, T> {
+        type Target = T;
+    }
+
+    struct Foo(u32);
+    impl Foo {
+        fn a<R: Receiver<Target=Self>>(self: R) { }
```

Comment:
Is that really something we want to allow? Is it ever useful?

Reply:
No, and it doesn't work. It's just that the current error message isn't very good, so this section is proposing we spot this case specifically and produce better diagnostics.

Reply:
> No, and it doesn't work.

Ah good. :) Receiver types should be concrete IMO, at least until we have a clear usecase.

Reply:
To be specific, the case which doesn't work is the fourth line of the `main` function in this example.

The third line *does* work - [here's a playground using the existing unstable arbitrary self types support](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=d014901f33cec65187e3c2be85e7a04e).

So, if we wanted to forbid generic `Receiver`s, we'd have to _actively block it_. I'm fine with that if the consensus is that it's a good idea. I haven't heard of any use-cases for generic receivers. But I think we'll need to add a justification to this RFC - is it fair to say that we should block this just to reduce unknown unknowns, or will we need something a bit more crisp? (I'd note that this has been possible with the unstable `arbitrary_self_types` support for years)

Reply:
> is it fair to say that we should block this just to reduce unknown unknowns,

Yeah that seems like a reasonable argument to me. There is a risk of subtle interactions with method resolutions and other parts of the language when we can't even know the receiver type of a method, and we don't know a usecase, so there's no reason to work out the details of what exactly happens in this case.

Reply:
OK, fad04aee amends this RFC to add a proposal that we ban generic receiver types. If that's OK with you I'll resolve this conversation tomorrow.